### PR TITLE
NEW|PASINDU|CODE UPDATE

### PR DESCRIPTION
**Pull request by PasinduUkwatta**

### CODE REVIEW

**Review by shafraz17 — DISMISSED** [review]

File: client/BootstrapClient.java [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
     public static void main(String[] args) {
         Scanner scanner = new Scanner(System.in);
-        try {
-            DatagramSocket socket = new DatagramSocket();
-            InetAddress address = InetAddress.getByName("localhost"); // Change to server's address if needed
+        try (DatagramSocket socket = new DatagramSocket();
+             FileWriter fileWriter = new FileWriter(LOG_FILE, true); // Append mode
+             PrintWriter logWriter = new PrintWriter(fileWriter)) {
+
+            InetAddress address = InetAddress.getByName("localhost");
```
Comment (hasinthaK) [inline client/BootstrapClient.java]: This should be reverted, as I mentioned in an earlier PR as well, @shafraz17 & myself agreed to have the server's static IP hardcoded in here. However, localhost is totally incorrect. 
please check previous version & add the server's static public IP here.

**Review by hasinthaK — CHANGES_REQUESTED** [review]
Check the server name "localhost" & revert it.
Thanks

File: server/BootstrapServer.java [hunk 2]
```diff
@@ -177,23 +203,39 @@ public static void main(String args[])
...
+                    boolean found = false;
                     for (int i=0; i<nodes.size(); i++) {
                         if (nodes.get(i).getPort() == port) {
                             nodes.remove(i);
-                            String reply = "0012 UNROK 0";
+                            reply = "0012 UNROK 0";
                             DatagramPacket dpReply = new DatagramPacket(reply.getBytes() , reply.getBytes().length , incoming.getAddress() , incoming.getPort());
                             sock.send(dpReply);
+                            found = true;
+                            break;
```
Comment (hasinthaK) [inline server/BootstrapServer.java:214]: hope this is intentional?
why does this show as a change?

File: server/BootstrapServer.java [hunk 3]
```diff
@@ -177,23 +203,39 @@ public static void main(String args[])
                     String ip = st.nextToken();
                     int port = Integer.parseInt(st.nextToken());
                     String username = st.nextToken();
+                    boolean found = false;
                     for (int i=0; i<nodes.size(); i++) {
                         if (nodes.get(i).getPort() == port) {
```
Comment (hasinthaK) [inline server/BootstrapServer.java:208]: @shafraz17 @PasinduUkwatta Are we checking only the port for the node connectivity? this seems incorrect,
two nodes might communicate using the same port, only the `ip:port` combination should be unique for a node.

OR

Am I missing something?

**Review by hasinthaK — COMMENTED** [review]
Minor clarification comments added. Please review them before merging.
No changes are requested from me.

Add updates for the comments if necessary please.

File: server/BootstrapServer.java (hunk 3 above)
Comment (shafraz17) [reply server/BootstrapServer.java:208]: This was the provided server code. This is related to the UNREG command right. according to the code, it only checks for the port right when removing a node

Comment (hasinthaK) [reply server/BootstrapServer.java:208]: if this part is provided then we dont want to worry, however we need to ensure that this does not affect impl.

File: server/BootstrapServer.java (hunk 2 above)
Comment (shafraz17) [reply server/BootstrapServer.java:214]: @PasinduUkwatta do we need this flag here? Breaking the loop is it a good idea?

**Review by shafraz17 — CHANGES_REQUESTED** [review]

File: server/BootstrapServer.java (hunk 2 above)
Comment (PasinduUkwatta) [reply server/BootstrapServer.java:214]: I added the flag to check whether the selected node is available or not, the loop break added due to stop the for loop after find the node. Else the loop will continue until whole node  size.

**Review by shafraz17 — APPROVED** [review]

**Review by hasinthaK — APPROVED** [review]

**Pull request merged**